core/src/main/java/org/infinispan/cache/impl/EncoderCache.java

```diff
@@ -0,0 +1,878 @@
+package org.infinispan.cache.impl;
+
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import javax.security.auth.Subject;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.Cache;
+import org.infinispan.CacheCollection;
+import org.infinispan.CacheSet;
+import org.infinispan.CacheStream;
+import org.infinispan.commands.read.AbstractCloseableIteratorCollection;
+import org.infinispan.commons.dataconversion.ByteArrayWrapper;
+import org.infinispan.commons.dataconversion.Encoder;
+import org.infinispan.commons.dataconversion.Wrapper;
+import org.infinispan.commons.util.CloseableIterator;
+import org.infinispan.commons.util.CloseableIteratorMapper;
+import org.infinispan.commons.util.CloseableSpliterator;
+import org.infinispan.commons.util.CloseableSpliteratorMapper;
+import org.infinispan.commons.util.InjectiveFunction;
+import org.infinispan.container.InternalEntryFactory;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.ForwardingCacheEntry;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.context.Flag;
+import org.infinispan.factories.ComponentRegistry;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.metadata.Metadata;
+import org.infinispan.notifications.cachelistener.ListenerHolder;
+import org.infinispan.notifications.cachelistener.filter.CacheEventConverter;
+import org.infinispan.notifications.cachelistener.filter.CacheEventFilter;
+
+/**
+ * Cache decoration that can transcode K/V based on requested MimeType.
+ * @since 9.1
+ */
+public class EncoderCache<K, V> extends AbstractDelegatingAdvancedCache<K, V> {
+
+   private InternalEntryFactory entryFactory;
+
+   private final Function<V, V> transcodeValueForRead = (InjectiveFunction & Function<V, V>) this::valueFromStorage;
```

Comment:
Why don't you just go ahead and change `InjectiveFunction` to extends `Function<T, R>` then we don't have to worry about these casts all over the place. Not sure why I didn't do this originally, sorry.